blueocean-events/src/main/java/io/jenkins/blueocean/events/PipelineEventListener.java

```diff
@@ -58,77 +57,63 @@
                 branch.add(flowNode.getId());
-                publishEvent(newMessage(PipelineEventChannel.Event.pipeline_block_start, flowNode, branch));
+                publishEvent(newMessage(PipelineEventChannel.Event.pipeline_block_start, flowNode, branch.iterator()));
             } else if (flowNode.getPersistentAction(QueueItemAction.class) != null) {
                 // Make sure we fire an event for the start of node blocks.
-                List<String> branch = getBranch(flowNode);
-                publishEvent(newMessage(PipelineEventChannel.Event.pipeline_step, flowNode, branch));
+                ArrayDeque<String> branch = getBranch(flowNode);
+                publishEvent(newMessage(PipelineEventChannel.Event.pipeline_step, flowNode, branch.iterator()));
             }
         } else if (flowNode instanceof StepAtomNode) {
-            List<String> branch = getBranch(flowNode);
-            publishEvent(newMessage(PipelineEventChannel.Event.pipeline_step, flowNode, branch));
+            ArrayDeque<String> branch = getBranch(flowNode);
+            publishEvent(newMessage(PipelineEventChannel.Event.pipeline_step, flowNode, branch.iterator()));
         } else if (flowNode instanceof StepEndNode) {
             if (flowNode.getAction(BodyInvocationAction.class) != null) {
                 FlowNode startNode = ((StepEndNode) flowNode).getStartNode();
                 String startNodeId = startNode.getId();
 
-                List<String> branch = getBranch(startNode);
+                ArrayDeque<String> branch = getBranch(startNode);
 
                 branch.add(startNodeId);
-                publishEvent(newMessage(PipelineEventChannel.Event.pipeline_block_end, flowNode, branch));
+                publishEvent(newMessage(PipelineEventChannel.Event.pipeline_block_end, flowNode, branch.iterator()));
             }
         } else if (flowNode instanceof FlowEndNode) {
             publishEvent(newMessage(PipelineEventChannel.Event.pipeline_end, flowNode.getExecution()));
         }
     }
 
-    private List<String> getBranch(FlowNode flowNode) {
-        List<String> branch = new ArrayList<>();
-        FlowNode parentBlock = getParentBlock(flowNode);
-
-        while (parentBlock != null) {
-            branch.add(0, parentBlock.getId());
-            parentBlock = getParentBlock(parentBlock);
-        }
-
-        return branch;
-    }
-
-    private FlowNode getParentBlock(FlowNode flowNode) {
-        List<FlowNode> parents = flowNode.getParents();
+    private ArrayDeque<String> getBranch(FlowNode flowNode) {
+        ArrayDeque<String> branch = new ArrayDeque<>();
+        String parentBlockId = flowNode.getEnclosingId();
 
-        for (FlowNode parent : parents) {
-            if (parent instanceof StepStartNode) {
-                if (parent.getAction(BodyInvocationAction.class) != null) {
-                    return parent;
+        while (parentBlockId != null) {
```

Comment:
I think what you want is `myFlowNode.getEnclosingBlocks()` which returns a list of blocks -- perhaps combined with Guava or Java 8 stream/filter APIs, perhaps via  `new NodeStepTypePredicate('parallel'))` or `FlowScanningUtils.hasActionPredicate(SomeAction.class))` -- it'll do all these goodies for you.

Reply:
@svanoort I looked at `myFlowNode.getEnclosingBlocks()`, it doesn't return blocks in order. This code was not written to me and I want to keep changes to minimum as it might break UI. Previous code was making sure preserving certain order, so I would prefer leaving the code with  fixing SOF problem with minor enhancements.

Reply:
@vivek Could we just do a `Lists.reverse`, to maintain the order? It's iterating in inside-out order here so I think you want to simply invert that.

Reply:
@svanoort Right. I was not planning to use was due to this javadoc of `GraphLookupView.findAllEnclosingBlockStarts() ` , https://github.com/jenkinsci/workflow-api-plugin/blob/a20f21fddc6831f51c3ceeefd3ed1b744b899e98/src/main/java/org/jenkinsci/plugins/workflow/graph/GraphLookupView.java#L55 that says order is not guaranteed. After clarification offline with you, I am fine using it. Please open  a PR to fix the Javadoc to make it clear that inside-out order will be maintained.